x-pack/auditbeat/module/system/package/package.go

```diff
@@ -0,0 +1,537 @@
+
+	bucketName              = "package.v1"
+	bucketKeyPackages       = "packages"
+	bucketKeyStateTimestamp = "state_timestamp"
+
+	eventTypeState = "state"
+	eventTypeEvent = "event"
+)
+
+type eventAction uint8
+
+const (
+	eventActionExistingPackage eventAction = iota
+	eventActionPackageInstalled
+	eventActionPackageRemoved
+)
+
+func (action eventAction) String() string {
+	switch action {
+	case eventActionExistingPackage:
+		return "existing_package"
+	case eventActionPackageInstalled:
+		return "package_installed"
+	case eventActionPackageRemoved:
+		return "package_removed"
+	default:
+		return ""
+	}
+}
+
+func init() {
+	mb.Registry.MustAddMetricSet(moduleName, metricsetName, New,
+		mb.DefaultMetricSet(),
+		mb.WithNamespace(namespace),
+	)
+}
+
+// MetricSet collects data about the system's packages.
+type MetricSet struct {
+	mb.BaseMetricSet
+	config    config
+	log       *logp.Logger
+	cache     *cache.Cache
+	bucket    datastore.Bucket
+	lastState time.Time
+	osFamily  string
+}
+
+// Package represents information for a package.
+type Package struct {
+	Name        string
+	Version     string
+	Release     string
+	Arch        string
+	License     string
+	InstallTime time.Time
+	Size        uint64
+	Summary     string
+	URL         string
+}
+
+// Hash creates a hash for Package.
+func (pkg Package) Hash() uint64 {
+	h := xxhash.New64()
+	h.WriteString(pkg.Name)
+	h.WriteString(pkg.InstallTime.String())
+	return h.Sum64()
+}
+
+func (pkg Package) toMapStr() common.MapStr {
+	mapstr := common.MapStr{
+		"name":        pkg.Name,
+		"version":     pkg.Version,
+		"installtime": pkg.InstallTime,
+	}
+
+	if pkg.Arch != "" {
+		mapstr.Put("arch", pkg.Arch)
+	}
+
+	if pkg.License != "" {
+		mapstr.Put("license", pkg.License)
+	}
+
+	if pkg.Release != "" {
+		mapstr.Put("release", pkg.Release)
+	}
+
+	if pkg.Size != 0 {
+		mapstr.Put("size", pkg.Size)
+	}
+
+	if pkg.Summary != "" {
+		mapstr.Put("summary", pkg.Summary)
+	}
+
+	if pkg.URL != "" {
+		mapstr.Put("url", pkg.URL)
+	}
+
+	return mapstr
+}
+
+func getOS() (*types.OSInfo, error) {
+	host, err := sysinfo.Host()
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting the OS")
+	}
+
+	hostInfo := host.Info()
+	if hostInfo.OS == nil {
+		return nil, errors.New("no host info")
+	}
+
+	return hostInfo.OS, nil
+}
+
+// New constructs a new MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The %v/%v dataset is experimental", moduleName, metricsetName)
+
+	config := defaultConfig()
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack the %v/%v config", moduleName, metricsetName)
+	}
+
+	bucket, err := datastore.OpenBucket(bucketName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open persistent datastore")
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+		log:           logp.NewLogger(metricsetName),
```

Comment:
A what? A "metricset"? ;-)

Reply:
Indeed :) We might never be able to get it out of the whole codebase. For now at least everything that is user facing calls it "dataset".